src/backend/opencl/program.cpp

```diff
@@ -67,4 +69,32 @@ namespace opencl
             throw;
         }
     }
+
+    NDRange getKernelLaunchCfg1D(const Kernel& kernel)
+    {
+        int device = getActiveDeviceId();
+
+        size_t dim0ThreadCount = getDeviceWorkGroupSize(kernel, device);
+
+        // make sure you are not launching what the device can't handle
+        dim0ThreadCount = std::min(getDevice(device).getInfo<CL_DEVICE_MAX_WORK_GROUP_SIZE>(), dim0ThreadCount);
+
+        return cl::NDRange(dim0ThreadCount, 1, 1);
+    }
+
+    NDRange getKernelLaunchCfg2D(const Kernel& kernel)
+    {
+        int device = getActiveDeviceId();
+
+        size_t dim0ThreadCount = kernel.getWorkGroupInfo<CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE>(getDevice(device));
+
+        size_t recomWorkGroupSize = getDeviceWorkGroupSize(kernel, device);
+
+        size_t dim1ThreadCount = divup(recomWorkGroupSize, dim0ThreadCount);
+
+        // make sure you are not launching what the device can't handle
+        dim1ThreadCount = std::min(getDevice(device).getInfo<CL_DEVICE_MAX_WORK_GROUP_SIZE>(), dim1ThreadCount);
+
+        return cl::NDRange(dim0ThreadCount, dim1ThreadCount, 1);
+    }
```

Comment:
If the constant is defined in the kernel itself, it should be fine. If it is a compile time option (based on the threads) that is also being used for the launch configuration, it is going to be an issue (for example, checkout convolve). 

One solution would be to always have dynamically allocated `__local` memory in OpenCL, but this may not give the best performance.